sklearn/utils/tests/test_utils.py

```diff
@@ -225,10 +297,69 @@ def test_safe_indexing_pandas():
         assert_array_equal(np.array(X_df_indexed), X_indexed)
 
 
-def test_safe_indexing_mock_pandas():
+@pytest.mark.parametrize(
+    "X, key, err_msg",
+    [(X_toy, 1.0, "No valid specification of the columns."),
+     (X_toy, ['col_0'], "Specifying the columns using strings is only")]
+)
+def test_safe_indexing_axis_1_error(X, key, err_msg):
+    with pytest.raises(ValueError, match=err_msg):
+        safe_indexing(X, key, axis=1)
+
+
+@pytest.mark.parametrize("axis", [None, 3])
+def test_safe_indexing_error_axis(axis):
+    with pytest.raises(ValueError, match="'axis' should be either 0"):
+        safe_indexing(X_toy, [0, 1], axis=axis)
+
+
+@pytest.mark.parametrize("X_constructor", ['array', 'series'])
+def test_safe_indexing_1d_array_error(X_constructor):
+    # check that we are raising an error if the array-like passed is 1D and
+    # we try to index on the 2nd dimension
+    X = list(range(5))
+    if X_constructor == 'array':
+        X_constructor = np.asarray(X)
+    elif X_constructor == 'series':
+        pd = pytest.importorskip("pandas")
+        X_constructor = pd.Series(X)
+
+    err_msg = "'X' should be a 2D NumPy array, 2D sparse matrix or pandas"
+    with pytest.raises(ValueError, match=err_msg):
+        safe_indexing(X_constructor, [0, 1], axis=1)
+
+
+@pytest.mark.parametrize(
+    "key, err_msg",
+    [(10, r"all features must be in \[0, 2\]"),
+     ('whatever', 'A given feature is not a column of the dataframe')]
+)
+def test_get_column_indices_error(key, err_msg):
+    pd = pytest.importorskip("pandas")
+    X_df = pd.DataFrame(X_toy, columns=['col_0', 'col_1', 'col_2'])
+
+    with pytest.raises(ValueError, match=err_msg):
+        _get_column_indices(X_df, key)
+
+
+@pytest.mark.parametrize(
+    "idx",
+    [[0, 1],
+     [True, True, False, False]]
+)
+@pytest.mark.parametrize("asarray", [True, False], ids=["array-like", "array"])
+def test_safe_indexing_pandas_series(idx, asarray):
+    pd = pytest.importorskip("pandas")
+    idx = np.asarray(idx) if asarray and isinstance(idx, Iterable) else idx
```

Comment:
Do we need the `isinstance(idx, Iterable)` check?

Reply:
good catch